### PR TITLE
feat: add sync/async option to the Kafka producer bridge

### PR DESCRIPTION
**Pull request by kjellwinblad**
This commit makes it possible to configure if a Kafka bridge should work in query mode sync or async by setting the resource_opts.query_mode configuration option.

Fixes:
https://emqx.atlassian.net/browse/EMQX-8631

### <samp>🤖 Generated by Copilot at 374ed1f</samp>

This pull request adds more flexibility and control over the Kafka producer resource in the `emqx_bridge_kafka` application. It introduces new configuration options for the query mode and timeout of the resource, as well as the translation of the resource options fields. It also enhances the query functionality of the `emqx_resource` module for built-in buffer resources.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [x] Changed lines covered in coverage report
- [x] Change log has been added to `changes/{ce,ee}/(feat|perf|fix)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [x] If there should be document changes, a PR to emqx-docs.git is sent, or a jira ticket is created to follow up
- [x] Schema changes are backward compatible

### CODE REVIEW

File: apps/emqx_bridge_kafka/src/emqx_bridge_kafka.erl [hunk 1]
```diff
@@ -294,7 +294,37 @@ fields(producer_kafka_opts) ->
...
+fields(resource_opts_fields) ->
+    [
+        {query_mode,
+            mk(
+                enum([async, sync]),
+                #{
+                    default => async,
+                    desc => ?DESC(query_mode)
+                }
+            )}
```
Comment (zmstone) [inline apps/emqx_bridge_kafka/src/emqx_bridge_kafka.erl:327]:
```suggestion
            })},
        {query_mode,
            mk(
                enum([async, sync]),
                #{
                    default => async,
                    desc => ?DESC(query_mode)
                }
            )}，
        {sync_query_timeout,
            mk(
                emqx_schema:timeout_duration_ms(),
                #{
                    default => <<"5s">>,
                    desc => ?DESC(query_mode_sync_timeout)
                }
            )}
```

File: apps/emqx_bridge_kafka/src/emqx_bridge_kafka.erl [hunk 2]
```diff
@@ -294,7 +294,37 @@ fields(producer_kafka_opts) ->
...
+            mk(
+                emqx_schema:duration_ms(),
+                #{
+                    default => <<"5s">>,
+                    desc => ?DESC(query_mode_sync_timeout)
+                }
+            )}
+    ];
+fields(resource_opts) ->
+    [
```
Comment (zmstone) [inline apps/emqx_bridge_kafka/src/emqx_bridge_kafka.erl:308]:
```suggestion
            )},
```

File: apps/emqx_bridge_kafka/src/emqx_bridge_kafka.erl [hunk 3]
```diff
@@ -294,7 +294,37 @@ fields(producer_kafka_opts) ->
             mk(ref(producer_buffer), #{
                 required => false,
                 desc => ?DESC(producer_buffer)
-            })}
+            })},
+        {query_mode_sync_timeout,
```
Comment (zmstone) [inline apps/emqx_bridge_kafka/src/emqx_bridge_kafka.erl:298]: rename to `sync_query_timeout`, and move it to the `resource_opts_fields` struct.

Comment (kjellwinblad) [reply apps/emqx_bridge_kafka/src/emqx_bridge_kafka.erl]: I renamed the field but the `resource_opts_fields` struct does not exist after the refactoring I did based on your suggestion. Please let me know if you still think it needs to be changed.

File: apps/emqx_bridge_pulsar/src/emqx_bridge_pulsar_impl_producer.erl [hunk 4]
```diff
@@ -70,10 +70,9 @@
 
 callback_mode() -> async_if_possible.
 
-%% there are no queries to be made to this bridge, so we say that
-%% buffer is supported so we don't spawn unused resource buffer
-%% workers.
-is_buffer_supported() -> true.
+%% consumer bridges don't need resource workers
+query_mode(_Config) ->
+    no_queries.
```
Comment (zmstone) [inline apps/emqx_bridge_pulsar/src/emqx_bridge_pulsar_impl_producer.erl:75]: pulsar is like Kafka, it has a buffer.

Comment (zmstone) [reply apps/emqx_bridge_pulsar/src/emqx_bridge_pulsar_impl_producer.erl:75]: should return `simple_async` here to keep it compatible as before this PR.

Comment (kjellwinblad) [reply apps/emqx_bridge_pulsar/src/emqx_bridge_pulsar_impl_producer.erl:75]: Thanks, I think I was missled byt the comment that existed before my change.

File: apps/emqx_bridge_kafka/src/emqx_bridge_kafka.erl [hunk 5]
```diff
@@ -294,7 +294,23 @@ fields(producer_kafka_opts) ->
...
+            mk(
+                enum([async, sync]),
+                #{
+                    default => async,
+                    desc => ?DESC(query_mode)
+                }
+            )},
+        {sync_query_timeout,
+            mk(
+                emqx_schema:duration_ms(),
```
[the remark this answers is not in the document]
Comment (kjellwinblad) [reply apps/emqx_bridge_kafka/src/emqx_bridge_kafka.erl]: Thanks, fixed in latest commit.

**Comment by yanzhiemq** [comment]
_

**Comment by yanzhiemq** [comment]
_

**Pull request merged**